### DISCUSSION TITLE
Missing data on cognition.run

### DISCUSSION THREAD

**Discussion by veranicolette** [jspsych/jsPsych#1467] [Q&A]
Hello, I am running my experiment on cognition.run, but it is not saving my very last trial. I'm not sure what I'm doing wrong, I'm supposed to have 157 indices at the end, but it only collects 156. Could anyone help me with this? Thanks.

**Comment by becky-gilbert** [comment]
Maybe @javidalpe can help?
Also, are you redirecting participants to a different page at the end of the experiment?

**Reply by veranicolette** [reply]
It goes to a thank you page with a code that participants post to MTurk for verification that they completed the task, but it stays within the experiment.

**Comment by javidalpe** [comment]
Hello! It's Javier from cognition.run. I'd need more information about the experiment in order to help with the issue. May I know the code or the URL of the experiment?

**Reply by veranicolette** [reply]
Hi Javier! Here's the code. The part that is not saving is independent task 2. It contains three sentences that participants have to complete and we record their responses. They're all showing up when I run it, but the very last sentence isn't getting recorded.

`var timeline = [];

        jsPsych.data.addProperties({
            subject_id: jsPsych.randomization.randomID(15),
        });

        var consent = {
            type: "html-button-response",
            stimulus: '<iframe src = "consent/consent_form.pdf" style="width:1000px; height:800px;" frameborder="0" allowfullscreen></iframe> ' + '<br></br>',
            choices: ["I HAVE READ THE PRECEDING INFORMATION AND WOULD LIKE TO PARTICIPATE IN THE STUDY", "I WOULD LIKE TO WITHDRAW FROM THIS STUDY. (There is no penalty for withdrawing.)"],
            
            on_finish: function(data) {
                if (data.button_pressed == 1) {
                jsPsych.endExperiment("Thank you for your response. <p> Please return this HIT through the Amazon Mechanical Turk platform (you may close this browser window). </p>");
                }
            }
        };
        timeline.push(consent);

        var instructions = {
            type: "html-keyboard-response",
            stimulus: "<p> In this task, you will see the beginning of a sentence and you will finish the sentence with the first thing that comes to mind. </p> <p> Please press the space bar to start </p>",
            choices: [32],
        };
        timeline.push(instructions);

        var independent_task1 = {
            type: "survey-html-form",
            html: jsPsych.timelineVariable('stimuli'),   
        }

        var print_out_independent_task1 = {
            timeline: [independent_task1],
            timeline_variables: independent_task_1,
            randomize_order: true,
            data: {
                given_sentence: jsPsych.timelineVariable('stimuli')
            }
        };
        timeline.push(print_out_independent_task1);

        var count = 0;
        var n_trials = 150;

        var sentences;
        var version = jsPsych.randomization.sampleWithoutReplacement([1,2],1)[0];
        console.log("version: ", version);
        if (version == 1){
            sentences = stimuli_list1;
        } else {
            sentences = stimuli_list2;
        }

        var trial = {
            type: "survey-html-form",
            html: jsPsych.timelineVariable('SENTENCE'),
        }

        var print_out_sentences = {
            timeline: [trial],
            timeline_variables: sentences,
            randomize_order: true,
            data: {given_sentence: jsPsych.timelineVariable('SENTENCE'),
                    stimuli_version: version,
            },
            on_finish: function(data) {
                    count++;
                    var progress = count/n_trials;
                    jsPsych.setProgressBar(progress);
            }

        }
        timeline.push(print_out_sentences);

        var independent_task2 ={
            type: "survey-html-form",
            html: jsPsych.timelineVariable('stimuli'),
        }

        var print_out_independent_task2 = {
            timeline: [independent_task2],
            timeline_variables: independent_task_2,
            randomize_order: true,
            data: {
                    given_sentence: jsPsych.timelineVariable('stimuli')
            }
        };
        timeline.push(print_out_independent_task2);

        var code = {
            type: "html-keyboard-response",
            choices: jsPsych.NO_KEYS,
            stimulus: function(){
            return "You have completed the main task! Please copy the following code to the MTurk page. You can then close this window.  <br>" + jsPsych.data.get().values()[0].subject_id + "<br> Your browser may ask you to confirm that you want to to leave this page, just say leave, we have recorded your performance."
            }
        }
        timeline.push(code);

    
        jsPsych.init({
            timeline: timeline,
            show_progress_bar: true,
            auto_update_progress_bar: false,
        })`

**Comment by Patlane** [comment]
Hello!
I had the same issue today. Luckily I had two concluding trials at the end, so that I did not loose any data, but this is weird. I did my script exactly like Josh show in his youtube-video (https://www.youtube.com/watch?v=O628dNA7WCc&list=PLnfo1lBY1P2Mf_o6rV5wiqqn92Mw3UTGh&index=5&ab_channel=JoshdeLeeuw). I'm not a programmer, so I have no idea what to add to the experiment so that it would function better. What is really disturbing is that I redirect my participants to Prolific and cognition.run never gets a message that the run was actually finished. How could I set my experiment up such that cognition.run also knows that the run was really finished or not? Thank you so much!
